File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/BookDeleteTask.java
package org.nypl.simplified.books.controller;

import com.io7m.jfunctional.Unit;
import com.io7m.jnull.NullCheck;

import org.nypl.simplified.accounts.database.api.AccountType;
import org.nypl.simplified.books.api.Book;
import org.nypl.simplified.books.api.BookID;
import org.nypl.simplified.books.book_database.api.BookDatabaseEntryType;
import org.nypl.simplified.books.book_database.api.BookDatabaseException;
import org.nypl.simplified.books.book_database.api.BookDatabaseType;
import org.nypl.simplified.books.book_registry.BookRegistryType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.Callable;

final class BookDeleteTask implements Callable<Unit> {

  private static final Logger LOG = LoggerFactory.getLogger(BookDeleteTask.class);

  private final AccountType account;
  private final BookRegistryType book_registry;
  private final BookID book_id;

  BookDeleteTask(
      final AccountType account,
      final BookRegistryType book_registry,
      final BookID book_id) {

    this.account =
        NullCheck.notNull(account, "Account");
    this.book_registry =
        NullCheck.notNull(book_registry, "Book registry");
    this.book_id =
        NullCheck.notNull(book_id, "Book id");
  }

  @Override
  public Unit call() throws Exception {
    execute();
    return Unit.unit();
  }

  private void execute() throws BookDatabaseException {
    LOG.debug("[{}] deleting book", this.book_id.brief());

    final BookDatabaseType book_database = this.account.bookDatabase();
    final BookDatabaseEntryType entry = book_database.entry(this.book_id);
    entry.delete();
    final Book book = entry.getBook();
    this.book_registry.clearFor(book_id);
  }
}

File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/BookDeleteTask.kt
package org.nypl.simplified.books.controller

import org.nypl.simplified.accounts.api.AccountID
import org.nypl.simplified.accounts.database.api.AccountType
import org.nypl.simplified.books.api.BookID
import org.nypl.simplified.books.book_database.api.BookDatabaseException
import org.nypl.simplified.books.book_registry.BookRegistryType
import org.nypl.simplified.profiles.api.ProfileID
import org.nypl.simplified.profiles.api.ProfilesDatabaseType
import org.nypl.simplified.taskrecorder.api.TaskRecorder
import org.nypl.simplified.taskrecorder.api.TaskRecorderType
import org.nypl.simplified.taskrecorder.api.TaskResult
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class BookDeleteTask(
  accountID: AccountID,
  profileID: ProfileID,
  profiles: ProfilesDatabaseType,
  private val bookID: BookID,
  private val bookRegistry: BookRegistryType,
) : AbstractBookTask(accountID, profileID, profiles) {

  override val logger: Logger =
    LoggerFactory.getLogger(BookDeleteTask::class.java)

  override val taskRecorder: TaskRecorderType =
    TaskRecorder.create()

  @Throws(BookDatabaseException::class)
  override fun execute(account: AccountType): TaskResult.Success<Unit> {
    this.logger.debug("[{}] deleting book", this.bookID.brief())
    this.taskRecorder.beginNewStep("Deleting book...")

    return try {
      val entry = account.bookDatabase.entry(this.bookID)
      entry.delete()
      this.taskRecorder.finishSuccess(Unit)
    } catch (e: Exception) {
      this.taskRecorder.currentStepFailed(
        message = e.message ?: e.javaClass.canonicalName ?: "unknown",
        errorCode = "deleteFailed",
        exception = e
      )
      throw TaskFailedHandled(e)
    } finally {
      this.bookRegistry.clearFor(this.bookID)
    }
  }

  override fun onFailure(result: TaskResult.Failure<Unit>) {
    // Nothing to do
  }
}
